pygamelib/pygamelib.js

```diff
@@ -163,10 +164,117 @@ var PygameLib = {};
                 PygameLib.eventTimer[event].timer = setInterval(PygameLib.eventTimer[event].f, ms);
             }
+            return mod;
         });
+
+        mod.Clock = Sk.misceval.buildClass(mod, time_Clock, 'Clock', []);
+        PygameLib.ClockType = mod.Clock;
         return mod;
     };
 
+    function time_Clock($gbl, $loc) {
+        $loc.__init__ = new Sk.builtin.func(function (self) {
+            Sk.abstr.sattr(self, 'prevTime', Sk.builtin.none.none$, false);
+            Sk.abstr.sattr(self, 'getTime', Sk.builtin.none.none$, false);
+            Sk.abstr.sattr(self, 'rawTime', Sk.ffi.remapToPy(0), false);
+            Sk.abstr.sattr(self, 'fpsArray', Sk.ffi.remapToPy([]), false);
+            Sk.abstr.sattr(self, 'fpsIdx', Sk.ffi.remapToPy(0));
+            return Sk.builtin.none.none$;
+        }, $gbl);
+        $loc.__init__.co_name = new Sk.builtins['str']('__init__');
+
+        $loc.tick = new Sk.builtin.func(function (self, framerate) {
+            var currTime = Date.now();
+            var mills = 0;
+            if (Sk.ffi.remapToJs(Sk.abstr.gattr(self, 'prevTime', false)) !== null) {
+                var prevTime = Sk.ffi.remapToJs(Sk.abstr.gattr(self, 'prevTime', false));
+                mills = (currTime - prevTime);
+            }
+            Sk.abstr.sattr(self, 'prevTime', Sk.ffi.remapToPy(currTime), false);
+            Sk.abstr.sattr(self, 'getTime', Sk.ffi.remapToPy(mills), false);
+            var arr = Sk.ffi.remapToJs(Sk.abstr.gattr(self, 'fpsArray', false));
+            var idx = Sk.ffi.remapToJs(Sk.abstr.gattr(self, 'fpsIdx', false));
+            if (arr.length < 10) {
+                arr.push(mills);
+            } else {
+                arr[idx] = mills;
+            }
+            idx = (idx + 1) % 10;
+            Sk.abstr.sattr(self, 'fpsArray', Sk.ffi.remapToPy(arr), false);
+            Sk.abstr.sattr(self, 'fpsIdx', Sk.ffi.remapToPy(idx), false);
+            if (framerate !== undefined) {
+                var timeout = 1000 / Sk.ffi.remapToJs(framerate);
+                return new Sk.misceval.promiseToSuspension(
+                    new Promise(function(resolve) {
+                    var f = function() {
+                        Sk.abstr.sattr(self, 'rawTime', Sk.ffi.remapToPy(Date.now() - currTime), false);
+                        while (currTime + timeout >= Date.now()) {}
+                        resolve(mills);
+                    };
+                    Sk.setTimeout(f, 1);
+                }));
+            }
+            Sk.abstr.sattr(self, 'rawTime', Sk.ffi.remapToPy(Date.now() - currTime), false);
+            return Sk.ffi.remapToPy(mills);
+        }, $gbl);
+        $loc.tick.co_name = new Sk.builtins['str']('tick');
+        $loc.tick.co_varnames = ['framerate'];
+        $loc.tick.$defaults = [Sk.ffi.remapToPy(0)];
+
+        $loc.tick_busy_loop = new Sk.builtin.func(function (self, framerate) {
+            // TODO: Should this method have a different implementation?
```

Comment:
Methods tick() and tick_busy_loop() are fine with same implementation since we cannot control the CPU through JavaScript. What matters to us is user experience, not so much under the hood :)

Reply:
"while (currTime + timeout >= Date.now()) {}" treba izbegavati u svakom slučaju, a posebno ja JavaScript-u.  I ja sam za to da tick() i tick_busy_loop() implementiramo na isti način, ali tako da bude bez "busy loop". Kada gledam kod vidim da se vač koristi Sk.setTimeout(f, 1), dakle u setTimeout() može da se lepo sačeka koliko je vremena ostalo.

Što se tiše korisničkog iskustva, važno je da se browser ne koči, da se stranica brzo učitava, da što više funkcionalnosti radi, da ne puca, da lepo izgleda, da se tačno nacrta, a da li je savršeno precizan frama rate, to nije kritičan aspekt korisničkog iskustva.